src/Utilities/matrix_of_constraints.jl

```diff
@@ -0,0 +1,357 @@
+Base.:(==)(a::Box, b::Box) = a.lower == b.lower && a.upper == b.upper
+function Base.empty!(b::Box)
+    empty!(b.lower)
+    empty!(b.upper)
+    return b
+end
+function Base.resize!(b::Box, n)
+    resize!(b.lower, n)
+    resize!(b.upper, n)
+    return
+end
+function load_constants(b::Box{T}, offset, set) where {T}
+    flag = single_variable_flag(typeof(set))
+    b.lower[offset+1] = if iszero(flag & LOWER_BOUND_MASK)
+        typemin(T)
+    else
+        extract_lower_bound(set)
+    end
+    b.upper[offset+1] = if iszero(flag & UPPER_BOUND_MASK)
+        typemax(T)
+    else
+        extract_upper_bound(set)
+    end
+    return
+end
+
+"""
+    mutable struct MatrixOfConstraints{T,AT,BT,ST} <: MOI.ModelLike
+        coefficients::AT
+        constants::BT
+        sets::ST
+        are_indices_mapped::BitSet
+        caches::Union{Nothing, Vector}
+        function MatrixOfConstraints{T,AT,BT,ST}() where {T,AT,BT,ST}
+            return new{T,AT,BT,ST}(AT(), BT(), ST(), BitSet(), nothing)
+        end
+    end
+
+Represents affine constraints in a matrix form where the linear coefficients
+of the functions are stored in the `coefficients` field and the constants of the
+functions or sets are stored in the `sets` field. Additional information
+about the sets are stored in the `sets` field.
+
+This model can only be used as the `constraints` field of a
+`MOI.Utilities.AbstractModel`. When the constraints are added,
+they are stored in the `caches` field. They are only loaded in
+the `coefficients` and `constants` fields once `MOI.Utilities.final_touch`
+is called. For this reason, this should not be used with incremental
+building of the model but with a `MOI.copy_to` instead.
+
+The constraints can be added in two different ways:
+1) With `add_constraint` in which case a canonicalized copy
+   of the function is stored in `caches`.
+2) With `pass_nonvariable_constraints` in which case the functions and sets are
+   stored themselves in `caches` without mapping the variable indices.
+   The corresponding index in `caches` is added in `are_indices_mapped`.
+   This allows to avoid doing a copy of the function in case
+   the getter of `CanonicalConstraintFunction` does not make a copy
+   for the source model, e.g., this is the case of `VectorOfConstraints`.
+
+We illustrate this with an example. Suppose a model is copied from
+a `src::MOI.Utilities.Model` to a bridged model with a `MatrixOfConstraints`.
+For all the types that are not bridged, the constraints will be copied
+with `pass_nonvariable_constraints` hence the functions stored in
+`caches` are exactly the same as the ones stored in `src`.
+This is ok since this is only during the `copy_to` operation during which `src`
+cannot be modified.
+On the other hand, for the types that are bridged, the functions added
+may contain duplicates even if the functions did not contain duplicates in
+`src` so duplicates are removed with `MOI.Utilities.canonical`.
+"""
```

Comment:
I started making inline notes, but I'll stop. The API needed to implement MatrixOfConstraints is quite large, and needs to be documented.

This docstring, plus a page in the documentation, should mention what functions need to be implemented for the different `coefficients`, `constants`, and `sets` (e.g., `set_types`, `rows`, ...) fields.


See the comment at the top of this file, and the notes in `ProductOfSets.jl`